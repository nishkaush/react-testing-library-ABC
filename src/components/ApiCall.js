import React, { useState, useEffect } from "react";

const ApiCall = () => {
  const [todo, setTodo] = useState({});
  const [isLoading, setIsLoading] = useState(false);

  const makeApiCall = async () => {
    try {
      setIsLoading(true);
      const resp = await fetch("https://jsonplaceholder.typicode.com/posts/1");
      const data = await resp.json();
      setTodo(data);
      setIsLoading(false);
    } catch (error) {
      console.log("error", error);
      setIsLoading(false);
    }
  };

  return (
    <div>
      <h1>Api call</h1>
      {isLoading && <p>Loading...</p>}
      {!isLoading && (
        <>
          <p>
            Title-{todo.title} + Body-{todo.body}
          </p>
          <button onClick={makeApiCall} data-testid="my-btn">
            Get Todo
          </button>
        </>
      )}
    </div>
  );
};

export default ApiCall;
