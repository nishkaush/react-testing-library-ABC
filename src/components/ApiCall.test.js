import React from "react";
import {
  render,
  fireEvent,
  waitForElementToBeRemoved,
} from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";

import ApiCall from "./ApiCall";

fetch = jest.fn(() =>
  Promise.resolve({
    json: () => Promise.resolve({ title: "yahoo", body: "my body" }),
  })
);

it("renders", () => {
  const { asFragment } = render(<ApiCall />);
  expect(asFragment()).toMatchSnapshot();
});

it("shows loader when get-todo button is clicked", async () => {
  const { getByTestId, getByText, queryByText } = render(<ApiCall />);

  const getTodoBtn = getByTestId("my-btn");
  fireEvent.click(getTodoBtn);

  const loader = queryByText("Loading...");
  expect(loader).toBeInTheDocument();

  await waitForElementToBeRemoved(() => queryByText("Loading..."));
});
