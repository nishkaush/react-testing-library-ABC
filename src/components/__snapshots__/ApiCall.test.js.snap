// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
<DocumentFragment>
  <div>
    <h1>
      Api call
    </h1>
    <p>
      Title- + Body-
    </p>
    <button
      data-testid="my-btn"
    >
      Get Todo
    </button>
  </div>
</DocumentFragment>
`;
